### PR TITLE
Implementation of missing user management features

### LINKED ISSUE #1784: Implement missing user management features
* [x] Enable/disable local account creation (admin option)
* [x] Enable/disable notification to admins when new account is created
* [x] Enable/disable account moderation workflow
* [x] Creation of users

### PR DESCRIPTION
**Pull request by mic4ael**
fixes #1784

### CODE REVIEW

File: indico/modules/auth/controllers.py [hunk 1]
```diff
@@ -60,6 +61,7 @@ class RHLogin(RH):
 
     def _process(self):
         login_reason = session.pop('login_reason', None)
+        local_registration_enabled = HelperMaKaCInfo.getMaKaCInfoInstance().getAuthorisedAccountCreation()
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: MaKaCInfo is legacy/zodb/deprecated.
We can either put it in indico.conf or in settings (via `SettingsProxy`). Probably the latter makes more sense. In that case we need a really small zodbimporter module to copy  the existing data from zodb.
Check [indico_zodbimport/modules/api.py](https://github.com/indico/indico/blob/master/indico_zodbimport/modules/api.py) on a good example on how to do it.

File: indico/modules/auth/controllers.py [hunk 2]
```diff
@@ -117,18 +119,32 @@ def _process(self):
...
+                'the link in that email.'))
+    else:
+        flash(flashed_msg)
+
     return redirect(url_for(endpoint, **url_args))
 
 
+def _send_approval_mail(email):
+    url = url_for('.register', approval_token=secure_serializer.dumps(email, salt='approve-email-salt'), _external=True)
+    admin = db.session.query(User).filter_by(is_admin=True).first()
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: First of all, you'd use `User.find_first(is_admin=True)` or `User.query.filter_by(is_admin=True).first()` (the latter is more flask-sqlalchemy-style, the former more indico-style) instead of `db.session.query`

However, sending an email to the "first" admin is a terrible idea. Email all admins or add a new settings on who to notify about new registrations. I'd go for the setting since not all admins might care about it.

File: indico/modules/auth/controllers.py [hunk 3]
```diff
@@ -117,18 +119,32 @@ def _process(self):
...
     template_args = template_args or {}
     url_args = url_args or {}
     token = secure_serializer.dumps(data or email, salt=salt)
     url = url_for(endpoint, token=token, _external=True, _secure=True, **url_args)
     template_module = get_template_module(template, email=email, url=url, **template_args)
     GenericMailer.send(make_email(email, template=template_module))
-    flash(_('We have sent you a verification email. Please check your mailbox within the next hour and open '
-            'the link in that email.'))
+
+    if not flashed_msg:
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: I would always flash the msg in the function calling `_send_confirmation` instead of sometimes flashing a local msg and sometimes one from outside.

File: indico/modules/auth/controllers.py [hunk 4]
```diff
@@ -117,18 +119,32 @@ def _process(self):
...
+        flash(flashed_msg)
+
     return redirect(url_for(endpoint, **url_args))
 
 
+def _send_approval_mail(email):
+    url = url_for('.register', approval_token=secure_serializer.dumps(email, salt='approve-email-salt'), _external=True)
+    admin = db.session.query(User).filter_by(is_admin=True).first()
+    template = get_template_module('auth/emails/account_approval_email.txt', url=url, new_account_email=email, recipient=admin)
+    GenericMailer.send(make_email(admin.email, template=template))
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: use the `send_email()` util instead of `GenericMailer.send()`

File: indico/modules/auth/controllers.py [hunk 5]
```diff
@@ -204,6 +220,12 @@ class RHRegister(RH):
     - creation of a new user based on information from an identity provider
     """
 
+    def _checkProtection(self):
+        RH._checkProtection(self)
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: RH._checkProtection is always empty, no need to call it. also, we never link to that page if account creation is not active so i'd raise either `NotFound` or `Forbidden` (both can be imported from `werkzeug.exceptions`) instead of the custom `AccessError`.

File: indico/modules/auth/controllers.py [hunk 6]
```diff
@@ -224,9 +246,21 @@ def _get_verified_email(self):
             return None
         return secure_serializer.loads(request.args['token'], max_age=3600, salt='register-email')
 
+    def _get_approved_email(self):
+        if 'approval_token' not in request.args:
+            return None
+        return secure_serializer.loads(request.args['approval_token'], salt='approve-email-salt')
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: all the other places where we use `secure_serializer` don't add `-salt` in the salt string

File: indico/modules/auth/controllers.py [hunk 7]
```diff
@@ -243,7 +277,9 @@ def _process(self):
                                       User.all_emails.contains(db.func.any(list(handler.get_all_emails(form)))))
 
         if form.validate_on_submit():
-            if handler.must_verify_email:
+            if account_moderation_enabled and handler.must_verify_email:
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: We only want account moderation for local accounts. However, `must_verify_email` may also be true for remote providers (e.g. because we don't know that they send us a verified email). Check if `isinstance(handler, LocalRegistrationHandler)` instead. (or maybe add a `is_local` attribute to the registration handlers so you can check `handler.is_local`)

File: indico/modules/auth/controllers.py [hunk 8]
```diff
@@ -253,20 +289,30 @@ def _process(self):
...
+    def _send_confirmation(self, email, flashed_msg=None):
         session['register_verification_email_sent'] = True
         return _send_confirmation(email, 'register-email', '.register', 'auth/emails/register_verify_email.txt',
-                                  url_args={'provider': self.provider_name})
+                                  url_args={'provider': self.provider_name}, flashed_msg=flashed_msg)
 
     def _create_user(self, form, handler, pending_user):
         data = form.data
+
+        if session.get('approval_email_sent', False):
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: i think you could unconditionally delete it here..?

    session.pop('approval_email_sent', None)

File: indico/modules/auth/controllers.py [hunk 9]
```diff
@@ -301,8 +347,17 @@ def _create_user(self, form, handler, pending_user):
                 'Check <a href="{url}">your profile</a> for further details and settings.')
         flash(Markup(msg).format(url=url_for('users.user_profile')), 'success')
         db.session.flush()
+        self._notify_admins_if_necessary(user)
         return handler.redirect_success()
 
+    def _notify_admins_if_necessary(self, new_user):
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: bad function name. why not just `_notify_admins`?

File: indico/modules/auth/controllers.py [hunk 10]
```diff
@@ -301,8 +347,17 @@ def _create_user(self, form, handler, pending_user):
...
+        info = HelperMaKaCInfo.getMaKaCInfoInstance()
+        if info.getNotifyAccountCreation():
+            admins = db.session.query(User).filter_by(is_admin=True).all()
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: `User.find_all(is_admin=True)`

Also, send a single email to all of them. And I'd go for the separate email list setting I suggested above.

File: indico/modules/auth/controllers.py [hunk 11]
```diff
@@ -509,7 +564,8 @@ def create_form(self):
         return form
 
     def create_identity(self, data):
-        del session['register_verified_email']
+        if session.get('register_verified_email', False):
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: could be deleted unconditionally, too, i guess

File: indico/modules/auth/forms.py [hunk 12]
```diff
@@ -100,7 +100,7 @@ class MultipassRegistrationForm(SyncedInputsMixin, IndicoForm):
 
 
 class LocalRegistrationForm(RegistrationForm):
-    email = EmailField(_('Email address'))
+    email = EmailField(_('Email address'), [DataRequired(), _check_existing_email])
```
Comment (ThiefMaster) [inline indico/modules/auth/forms.py]: why? this field is always supposed to be disabled (and thus never in the submitted data) since we only get to that page after validating the email - so it's never editable.

File: indico/modules/auth/templates/emails/account_approval_email.txt [hunk 13]
```diff
@@ -0,0 +1,19 @@
+{% extends 'emails/base.txt' %}
+
+{% block subject -%}
+    {% trans %}Approval to send an email validation message{% endtrans %}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/account_approval_email.txt]: that sounds strange

File: indico/modules/auth/templates/emails/account_approval_email.txt [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    {{ recipient.full_name }}
+{%- endblock %}
+
+{% block body -%}
+    {% filter dedent %}
+        Click the link below to approve the request from address: {{ new_account_email }}
+        This approval will result in sending an email validation message to the user.
+
+        {{ url }}
+
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/account_approval_email.txt]: do we really want this linebreak?

File: indico/modules/auth/templates/emails/account_approval_email.txt [hunk 15]
```diff
@@ -0,0 +1,19 @@
+{% extends 'emails/base.txt' %}
+
+{% block subject -%}
+    {% trans %}Approval to send an email validation message{% endtrans %}
+{%- endblock %}
+
+{% block header_recipient -%}
+    {{ recipient.full_name }}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/account_approval_email.txt]: simply "admin" when sending a single email to multiple persons

File: indico/modules/auth/templates/emails/account_creation_email.txt [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    {% trans %}Account has been created{% endtrans %}
+{%- endblock %}
+
+{% block header_recipient -%}
+    {{ recipient.full_name }}
+{%- endblock %}
+
+{% block body -%}
+    {% filter dedent %}
+        A new Indico profile has been created for a user with name {{ user.full_name }} ({{ user.email }})
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/account_creation_email.txt]: I would show the information in new lines. "for a user with name" isn't very good english.

File: indico/modules/auth/templates/emails/account_creation_email.txt [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+        {{ url_for('users.user_dashboard', user_id=user.id, _external=True) }}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/emails/account_creation_email.txt]: in jinja the preferred way is `true`/`false`/`none` (all lowercase)

File: indico/modules/auth/templates/login_page.html [hunk 18]
```diff
@@ -62,18 +62,20 @@
...
-                                    {%- endtrans %}
-                                </p>
+                            {% if local_registration_enabled %}
+                                {% if indico_config.LocalIdentities %}
+                                    <p class="register">
+                                        {% trans url=url_for('.register', next=request.args.get('next')) -%}
+                                            If you do not have an Indico account yet, you can <a href="{{ url }}">create one here</a>.
+                                        {%- endtrans %}
+                                    </p>
+                                {% elif indico_config.ExternalRegistrationURL %}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/login_page.html]: This one should be visible even without local registration being enabled

File: indico/modules/users/controllers.py [hunk 19]
```diff
@@ -26,13 +26,16 @@
...
+from indico.core.config import Config
 from indico.core.notifications import make_email
 from indico.modules.users import User, logger
 from indico.modules.users.models.emails import UserEmail
 from indico.modules.users.util import (get_related_categories, get_suggested_categories,
                                        serialize_user, search_users, merge_users)
 from indico.modules.users.views import WPUserDashboard, WPUser, WPUsersAdmin
-from indico.modules.users.forms import UserDetailsForm, UserPreferencesForm, UserEmailsForm, SearchForm, MergeForm
+from indico.modules.users.forms import (UserDetailsForm, UserPreferencesForm,
+                                        UserEmailsForm, SearchForm, MergeForm, UserManagementForm)
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: at least two of those can go in the previous line

File: indico/modules/users/controllers.py [hunk 20]
```diff
@@ -50,10 +53,13 @@
 from MaKaC.common.cache import GenericCache
 from MaKaC.common.mail import GenericMailer
 from MaKaC.common.timezoneUtils import DisplayTZ
+from MaKaC.common.info import HelperMaKaCInfo
 from MaKaC.conference import CategoryManager
 from MaKaC.webinterface.rh.admins import RHAdminBase
 from MaKaC.webinterface.rh.base import RHProtected
 
+import transaction
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: this should be in the 3rd party imports, i.e. before the `from flask` import

File: indico/modules/users/controllers.py [hunk 21]
```diff
@@ -303,15 +312,69 @@ def _process(self):
...
+    def _load_management_settings(self):
+        info = HelperMaKaCInfo.getMaKaCInfoInstance()
+        return {
+            'notify_on_new_account': info.getNotifyAccountCreation(),
+            'local_account_creation': info.getAuthorisedAccountCreation(),
+            'account_moderation_workflow': info.getModerateAccountCreation()
+        }
+
+    def _user_management_form_posted(self):
+        return 'save_management_form' in self.getRequestParams()
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: getRequestParams is deprecated. use the `request.*` dicts from flask.

File: indico/modules/users/controllers.py [hunk 22]
```diff
@@ -303,15 +312,69 @@ def _process(self):
...
+        info.setAuthorisedAccountCreation(data.pop('local_account_creation'))
+        info.setModerateAccountCreation(data.pop('account_moderation_workflow'))
+        transaction.commit()
 
 
 class RHUsersAdminCreate(RHAdminBase):
     """Create user (admin)"""
 
     def _process(self):
+        # to avoid circular imports
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: that's pretty obvious when you have local imports ;)

File: indico/modules/users/controllers.py [hunk 23]
```diff
@@ -303,15 +312,69 @@ def _process(self):
...
+        from indico.modules.auth.controllers import LocalRegistrationHandler
+
         form = LocalRegistrationForm()
+        data = form.data
+        handler = LocalRegistrationHandler(self)
+
+        if form.validate_on_submit():
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: this block looke a LOT like duplicated code. can't we add a `create_user` util instead?

File: indico/modules/users/forms.py [hunk 24]
```diff
@@ -83,3 +83,16 @@ class MergeForm(IndicoForm):
...
+class UserManagementForm(IndicoForm):
+    notify_on_new_account = BooleanField(_('Notify when new account is created'), widget=SwitchWidget(),
+                                            description=_('Enable sending emails to admins when a new account is created'))
+    local_account_creation = BooleanField(_('Enable local registration'), widget=SwitchWidget(),
+                                            description=_('Enable users to register locally'))
+    account_moderation_workflow = BooleanField(_('Account moderation'), widget=SwitchWidget(),
+                                                description=_('Enable account moderation workflow'))
+
+    def is_submitted(self):
+        from flask import request
```
Comment (ThiefMaster) [inline indico/modules/users/forms.py]: why is this import local?

File: indico/modules/users/controllers.py (hunk 21 above)
Comment (ThiefMaster) [inline indico/modules/users/controllers.py]: also, why not use `is_submitted()` (which is also called by `validate_on_submit` btw) instead? you are overriding it after all...

File: indico/modules/users/templates/users_admin.html [hunk 25]
```diff
@@ -14,9 +14,20 @@
...
+                        <a class="i-button icon-plus" href="{{ url_for('.users_create') }}">Create a new account</a>
+                    </div>
                 </div>
             </div>
             <div class="i-box-content">
+                <h2>{% trans %} Account creation settings {% endtrans %} </h2>
+                {{ form_header(user_management_form) }}
+                {{ form_rows(user_management_form) }}
+                {% call form_footer(user_management_form) %}
+                    <input class="i-button big highlight" type="submit" value="{% trans %}Save{% endtrans %}" name="save_management_form">
```
Comment (ThiefMaster) [inline indico/modules/users/templates/users_admin.html]: `name="save_management_form"` - you could use a hidden input instead or even use a `SubmitButton` in the form definition itself (skip it in the `form_rows()` call and render it using `form.whatever(class_='i-button big highlight')`).

File: indico/modules/auth/controllers.py [hunk 26]
```diff
@@ -117,18 +119,32 @@ def _process(self):
         return multipass.logout(request.args.get('next') or url_for('misc.index'), clear_session=True)
 
 
-def _send_confirmation(email, salt, endpoint, template, template_args=None, url_args=None, data=None):
+def _send_confirmation(email, salt, endpoint, template, template_args=None, url_args=None, data=None, flashed_msg=None):
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: `flash_msg` or `custom_flash_msg` might fit better

**Comment by ThiefMaster** [comment]
When clicking the approval link we should check `session.user.is_admin`. You cannot assume that a link you send in an email is only accessed by the user - for example, malware scanners might check links in emails.

**Comment by ThiefMaster** [comment]
I wonder if sending the activation link by email is the best choice... Why not add a new model/table (not sure what the best name would be) `containing `id`, `email`,  `comment` and `approved`. When the user registers, we could actually perform the email validation first (the user would have the opportunity to enter a comment to the admins there) and then, instead of showing the registration page, add a record to the new table.

in the admin area we could then show a list of people who need to be approved and once you approve someone there we'd send an email to the approved email address with a new token (could be the same as the email verification token) telling the user he can proceed. 

he'd end up on the same page he used to verify his email, but the code there would check if the email has already been approved and thus let the user register instead.

whenever the registration is successful or the admin rejects the request, the record from the new table would be deleted.

File: indico/modules/auth/controllers.py (hunk 1 above)
Comment (mic4ael) [inline indico/modules/auth/controllers.py]: There is still a lot of code that uses `HelperMakacInfo`.

Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: Not a reason to keep it there :)

Comment (mic4ael) [inline indico/modules/auth/controllers.py]: Yeah, I will change it properly ;)

File: indico/modules/users/controllers.py (hunk 23 above)
Comment (mic4ael) [inline indico/modules/users/controllers.py]: Yeah, I thought about that and now I am sure it was a good idea ;)

File: indico/modules/users/forms.py (hunk 24 above)
Comment (mic4ael) [inline indico/modules/users/forms.py]: In order not to contaminate global namespace.

Comment (ThiefMaster) [inline indico/modules/users/forms.py]: Having imports global is the proper way to do it unless they have to be local to avoid circular imports. Python has a module scope so you don't pollute a truly global namespace like you would do e.g. in PHP.

Comment (mic4ael) [inline indico/modules/users/forms.py]: Sorry, by `global` I meant local to the module.

File: indico/modules/auth/controllers.py [hunk 27]
```diff
@@ -561,3 +550,32 @@ def _reset_password(self, identity):
...
+
+        # no validation needed so it is enough
+        # to check if it is a POST request
+        if form.is_submitted():
+            self._add_registration_request(session['register_verified_email'], form.comment.data)
+
+        return WPAuth.render_template('register.html', form=form, creating_registration_request=True)
+
+    def _add_registration_request(self, email, comment):
+        request = RegistrationRequest.find_first(email=email)
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py]: `request` is not a good name for this as it shadows `flask.request`

File: indico/modules/users/templates/users_admin.html [hunk 28]
```diff
@@ -14,9 +14,23 @@
...
+                    <div class="group">
+                        <a class="i-button icon-plus" href="{{ url_for('.users_create') }}">Create a new account</a>
+                    </div>
+                    <div class="group">
+                        <a class="i-button icon-info" href="{{ url_for('.registration_requests_list') }}">Registration requests</a>
+                    </div>
                 </div>
             </div>
             <div class="i-box-content">
+                <h2>{% trans %} Account creation settings {% endtrans %} </h2>
```
Comment (ThiefMaster) [inline indico/modules/users/templates/users_admin.html]: unnecessary whitespace

File: indico/util/emails.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def send_verification_email(email, template, salt, endpoint, data=None, url_args=None):
+    token = secure_serializer.dumps(data or email, salt=salt)
+    url = url_for(endpoint, email=email, token=token, _external=True, _secure=True)
+    template = get_template_module(template, email=email, url=url)
+    send_email(make_email(email, template=template))
+
+
+def notify_of_registration_request_approval(email, endpoint):
```
Comment (ThiefMaster) [inline indico/util/emails.py]: this doesn't belong in `indico.util`

File: indico/util/emails.py [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+    approval_token = secure_serializer.dumps(email)
+    url = url_for(endpoint, _external=True, approval_token=approval_token, _secure=True)
+    template = get_template_module('auth/emails/registration_request_approved.txt', url=url)
+    send_email(make_email(email, template=template))
+
+
+def send_notification_to_admins(template):
```
Comment (ThiefMaster) [inline indico/util/emails.py]: code in `indico.util` shouldn't import models on the top level; that's like asking for import issues later ;)
